examples/hello/bpReader/helloBPReader.cpp

```diff
@@ -22,16 +23,25 @@
 
 int main(int argc, char *argv[])
 {
+    // Test if the file has been written by hello_bpWriter:
+    std::string filename = "myVector_cpp.bp";
+    std::ifstream f(filename.c_str());
+    if (!f.good())
+    {
+       std::cerr << "The file " << filename << " does not exist."
```

Comment:
Since ADIOS2 is already throwing an exception of type `std::ios_base_failure` I'd move this to a catch block at the end.

Reply:
I agree with @williamfgc here on letting the ADIOS error handling already in place address this.  However, a useful change on this would would be to have the catch blocks guard their output to only occur on MPI rank 0.

Reply:
@williamfgc : Great point, done.

@chuckatkins : Try as I might, I couldn't get the error message to be emitted from rank != 0. Does `bpIO.Open` restrict to rank 0 under the hood?

Reply:
@NAThompson your assumption is correct, rank 0 will be the one trying to do an actual request to the file system at Open for the metadata file, to then broadcast its contents using MPI. The latter minimizes the request to the file system to 1.

Reply:
In that sense, there should be an `MPI_Finalize` inside the catch as it should finish all processes.

Reply:
@williamfgc : I tried putting `MPI_Finalize()` inside the catch, but there was no difference in behavior.

Reply:
@NAThompson the behavior should be the same, I think what @chuckatkins is asking is to emit the std::cerr message only from rank 0.  `if(rank == 0) { std::cerr << ... }`

Reply:
@williamfgc : Is there a possibility that `e.what()` gives a different message based on rank?

Reply:
In general, inside adios2 yes, there is always rank 0 ( or serial runs), rank > 0 is not always guarantee. For this particular case, and many, the rank doesn't matter much as it is not adding any useful info to the user to go and debug.

Reply:
> @williamfgc : I tried putting `MPI_Finalize()` inside the catch, but there was no difference in behavior.

I think what you may want here it `MPI_Abort()` here.

More generally, I would argue you identified a bug. When something goes wrong during opening a BP3 file, that usually is only identified on rank 0, so rank 0 will throw. But the other ranks not only don't notify the application of the error, I believe they will actually hang in the `MPI_Bcast` that  is used to distribute the meta data, because rank 0 won't participate as it has exited early when the exception was thrown. So the application has no way of recovering from on error during Open, as it will hang the library when run in parallel. (That's for BP3 -- other engines may act differently).

Reply:
@germasch : I've reproduced the hang you've predicted. Testing your proposed fix.